### PR TITLE
0072-feel-in

### PR DESCRIPTION
**Pull request by StrayAlien**

### CODE REVIEW

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 1]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>number not in: e1 in >e2</description>
+        <resultNode name="number_006_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_007_a">
+        <description>number in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, 
either the form `number not in:` shall be used,
either the simpler form `number:` would avoid misunderstanding as checking this description

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 2]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="number_007_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_007_b">
+        <description>number in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, 
either the form `number not in:` shall be used,
either the simpler form `number:` would avoid misunderstanding as checking this description

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 3]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="number_007_c" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_007_d">
+        <description>number in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, 
either the form `number not in:` shall be used,
either the simpler form `number:` would avoid misunderstanding as checking this description

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 4]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="number_007_d" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_007_e">
+        <description>number in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, 
either the form `number not in:` shall be used,
either the simpler form `number:` would avoid misunderstanding as checking this description

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 5]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="number_007_e" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_008_a">
+        <description>number in: e1 in (e2..e3]</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, 
either the form `number not in:` shall be used,
either the simpler form `number:` would avoid misunderstanding as checking this description

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 6]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="number_008_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_008_b">
+        <description>number in: e1 in (e2..e3]</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, 
either the form `number not in:` shall be used,
either the simpler form `number:` would avoid misunderstanding as checking this description

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 7]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="number_008_d" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_008_e">
+        <description>number in: e1 in (e2..e3]</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, 
either the form `number not in:` shall be used,
either the simpler form `number:` would avoid misunderstanding as checking this description

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 8]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="number_008_e" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="number_009_a">
+        <description>number: e1 in [e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: This form is more generic and less prone imho to misunderstanding :+1:

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 9]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>string not in: e1 in >e2</description>
+        <resultNode name="string_006_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="string_007_a">
+        <description>string in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I would simplify in the form `string:` or otherwise analogous comment as above.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 10]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="string_007_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="string_007_b">
+        <description>string in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I would simplify in the form `string:` or otherwise analogous comment as above.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 11]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="string_007_c" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="string_007_d">
+        <description>string in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I would simplify in the form `string:` or otherwise analogous comment as above.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 12]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="string_007_d" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="string_007_e">
+        <description>string in: e1 in (e2..e3)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I would simplify in the form `string:` or otherwise analogous comment as above.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 13]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="string_007_e" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="string_008_a">
+        <description>string in: e1 in (e2..e3]</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I would simplify in the form `string:` or otherwise analogous comment as above.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 14]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="string_008_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="string_008_b">
+        <description>string in: e1 in (e2..e3]</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I would simplify in the form `string:` or otherwise analogous comment as above.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 15]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="string_008_d" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="string_008_e">
+        <description>string in: e1 in (e2..e3]</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I would simplify in the form `string:` or otherwise analogous comment as above.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 16]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="string_014_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+
+    <testCase id="boolean_001">
+        <description>number in: e1 in [e2,e3,…] (endpoints)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `boolean:` not number.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 17]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="boolean_001" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="boolean_001_a">
+        <description>number not in: e1 in [e2,e3,…] (endpoints)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `boolean:` not number.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 18]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>date in: e1 in [e2,e3,…] (endpoints)</description>
+        <resultNode name="date_001" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="date_001_a">
+        <description>string not in: e1 in [e2,e3,…] (endpoints)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `date:` not string.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 19]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>date in: e1 in [e2,e3,…] (ranges)</description>
+        <resultNode name="date_002" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="date_002_a">
+        <description>string not in: e1 in [e2,e3,…] (ranges)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `date:` not string.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 20]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>date in: e1 in &lt;=e2</description>
+        <resultNode name="date_003_1" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="date_003_a">
+        <description>string not in: e1 in &lt;=e2</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `date:` not string.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 21]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>date in: e1 in &lt;e2</description>
+        <resultNode name="date_004" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="date_004_a">
+        <description>string not in: e1 in &lt;e2</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `date:` not string.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 22]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>date in: e1 in >=e2</description>
+        <resultNode name="date_005_1" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="date_005_a">
+        <description>string not in: e1 in >=e2</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `date:` not string.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 23]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>date in: e1 in >e2</description>
+        <resultNode name="date_006" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="date_006_a">
+        <description>string not in: e1 in >e2</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `date:` not string.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 24]
```diff
@@ -0,0 +1,2735 @@
...
+        <description>dateTime: e1 in unary tests</description>
+        <resultNode name="dateTime_014_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="list_001">
+        <description>number in: e1 in [e2,e3,…] (endpoints)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not number.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 25]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="list_001" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="list_001_a">
+        <description>number not in: e1 in [e2,e3,…] (endpoints)</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not number.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 26]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="list_001_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="list_011">
+        <description>boolean: e1 in e2</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not boolean.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 27]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="list_011" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="list_011_a">
+        <description>boolean: e1 in e2</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not boolean.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 28]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="list_011_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id="list_011_b">
+        <description>boolean: e1 in e2</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not boolean.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 29]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="list_013_a" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">false</value>
+            </expected>
+        </resultNode>
+    </testCase>
+-->
+
+    <testCase id="list_014">
+        <description>boolean: e1 in unary tests</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not boolean.

File: TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml [hunk 30]
```diff
@@ -0,0 +1,2735 @@
...
+        <resultNode name="list_014" type="decision">
+            <expected>
+                <value xsi:type="xsd:boolean">true</value>
+            </expected>
+        </resultNode>
+    </testCase>
+
+    <testCase id
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not boolean.

Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `list:` not boolean.

Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `context:` not number.

Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `context:` not number.

Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `context:` not boolean.

Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `context:` not boolean.

Comment (tarilabs) [inline TestCases/compliance-level-3/0072-feel-in/0072-feel-in-test-01.xml]: I believe typo, `context:` not boolean.

**Comment by StrayAlien** [comment]
Thanks @tarilabs. I'll sort it out.

**Comment by StrayAlien** [comment]
@tarilabs.  done - thanks.  Hopefully that looks better.

**Comment by StrayAlien** [comment]
@tarilabs I would like to revisit a requested change to this test (from here: https://github.com/dmn-tck/tck/pull/204#pullrequestreview-194604731):

```XML
    <decision name="list_014_b" id="_list_014_b">
        <variable name="list_014_b"/>
        <literalExpression>
            <text>[1,2,3] in ([[1,2,3,4]], [[1,2,3]])</text>
        </literalExpression>
    </decision>
```

With the change to the assertion that the result is to be `false`.  I originally had it as `true`.  Looking at it again, I think that it should be `true`.

To my mind, the 'in' when testing against a unary tests is really the same thing as a inputEntry being tested against in a decision table - the inputEntries are unary tests.  And I think the spec is pretty explicit here:

image
 
item b.  Reading that, the above test should be `true` as one of the unary tests values is a list, and the tested-for value is contained in that list.  Recalling now , this was the original intention of the test.
 
 comments welcome.

**Comment by StrayAlien** [comment]
@tarilabs @etirelli added a test to assert for `false` for the ambiguity mentioned here: https://github.com/dmn-tck/tck/pull/204#discussion_r249643220.  

```XML
    <decision name="list_015" id="_list_015">
        <variable name="list_015"/>
        <literalExpression>
            <text>[1,2,3] in ([1,2,3])</text>
        </literalExpression>
    </decision>
```

where `[1, 2, 3] in ([1, 2, 3])` should not be regarded as having a RHS unary test but as `[1, 2, 3] in <expression>`,  Actually makes it the same as `[1, 2, 3] in [1, 2, 3]` which is also asserted to be `false`. 

I did have my code treating it as a unary test as (IMO), that is what the grammar says - parentheses as RHS of an 'in' contains 'positive unary tests' .. which can also be a single test:

image

If there was a line above 51.d reading `expression , "in" , expression ;`  Then I think it would be `false` (as this test asserts) .... but I have a feeling it should actually be `true` ....
 
However .. at any rate, better to have a test asserting we're all doing the same thing.

Comments welcome.

Greg.

**Comment by opatrascoiu** [comment]
I think the expected value for

```
    <decision name="list_011" id="_list_011">
        <variable name="list_011"/>
        <literalExpression>
            <text>[1,2,3] in [1,2,3]</text>
        </literalExpression>
    </decision>
```
and 

```
    <decision name="list_014_b" id="_list_014_b">
        <variable name="list_014_b"/>
        <literalExpression>
            <text>[1,2,3] in ([[1,2,3,4]], [[1,2,3]])</text>
        </literalExpression>
    </decision>
```

is debatable.  I think is should be true. 

I also think we should review the above semantics & Table 52.

**Comment by tarilabs** [comment]
I believe https://github.com/dmn-tck/tck/pull/228#issuecomment-462098982 applies to Decision Table semantics; I am not sure I understand the connection with Literal Expression.

I believe https://github.com/dmn-tck/tck/pull/228#issuecomment-462101934 is an edge case of ambiguity as originally discussed https://github.com/dmn-tck/tck/pull/204#discussion_r249643220 hence imho is not wise to include a test until we have a. an OMG RTF issue raised b. said issue addressed.

**Comment by agilepro** [comment]
The last change ends up testing soem edge cases.  Matteo feels if the last change was reverted the tests would be OK for inclusion.

**Comment by tarilabs** [comment]
One is an edge case, the other I indeed I don't agree with the value-result change ;)

**Comment by StrayAlien** [comment]
@agilepro @tarilabs - re the ambiguity test from comment https://github.com/dmn-tck/tck/pull/228#issuecomment-462101934.  Personally, I think we need tests like these.  The treatment of lists and unary test(s) in and out of decision tables is a minefield - I am happy I haven't had to deal with singleton type equality!!  

I can comment the test out if we agree - but I'd rather we decided upon an accepted behaviour and kept a test for it.

<said smiling> So far I am feeling happy I haven't held my breath waiting for clarification via RTF.  :-)  

Re the change of value result for test `list_014_b`.  I really do think this should be the same in a DT and in a literal expression.  Otherwise we'll have the situation where exactly the same thing is `false` in one place and `true` in another - eg in a literal expression:

`[1,2,3] in ([[1,2,3,4]], [[1,2,3]])`

is `false` (as it has been changed to),

but in a DT, with an input value of `[1,2,3]` (the same test value) and a rule input entry of (the same unary tests): 

`[[1,2,3,4]], [[1,2,3]]` 

the result would be `true`.  That is the same value being tested for against the same unary tests.  I really feel it should be consistent and I believe in both cases it should be `true`.

Note also the paragraph above table 50:

> FEEL supports additional syntactic sugar for comparison. Note that Grammar Rules (clause 10.3.1.2) are used in decision table condition cells. These decision table syntaxes are defined in Table 50.

So, given the grammar is to be used in DT Cells, and the 'in' syntax is defined for DTs.  I believe they should produce the same result.    

I really mean no offense @tarilabs , but I have reverted the test result value back to how it was in the original PR consistent with what I feel the result should be - I think what is what the PR should reflect until we sort it out.

**Comment by StrayAlien** [comment]
@opatrascoiu.  The reason `list_011` is false for `[1,2,3] in [1,2,3]`, albeit weird-ish as it is, is these two rules in table 52:

image

**Comment by tarilabs** [comment]
In comment https://github.com/dmn-tck/tck/pull/228#issuecomment-462135835
a reference is made multiple times to "Table 52" but I am struggling to see how does it connect to this test suite.
image

On other hand, I do concur --as proved by this thread-- that the result is debatable :)

**Comment by StrayAlien** [comment]
@tarilabs Sorry!  Table 50! Funny, because after @opatrascoiu said 52 I even checked the v1.1 spec to see if table 52 was the 'in' table there - but nope - then I went and did the same thing myself!  doh!!

I have edited that last post.  Thanks.  👍 

> On other hand, I do concur --as proved by this thread-- that the result is debatable :)

And that made me laugh. :-)

**Comment by agilepro** [comment]
There are a couple of tests which members disagree on the answer.   We should split this into two parts: the less controversial ones, and save the more controversial ones for more scrutiny.   Can someone comment on which ones are obvious and which ones controversial?

**Comment by agilepro** [comment]
Still looking for someone to designate the tests that are acceptable today.

**Comment by StrayAlien** [comment]
Not sure if we've raised a clarification on this as yet.

If not, I am happy to.

**Comment by agilepro** [comment]
Greg,

Matteo says there are two tests that are questionable in this set.   There are a large number of tests, so I would like to see the rest of them checked in, with only those two removed.

Test "list014-b"   
Test "list015"

Just remove those two to a different PR, and we can discuss them there.   The rest should PASS on RedHat and that is good for checking in.

**Comment by StrayAlien** [comment]
@agilepro @tarilabs - thanks.

I have removed both those tests for now.  I'll add a new PR with them back in after this get merged.

Thanks again.

Greg.

**Comment by agilepro** [comment]
I added an "approved" label on this so we can remember to commit this at the beginning of next month.

**Comment by opatrascoiu** [comment]
@StrayAlien @tarilabs 

I looked briefly at the test cases. I am not entirely sure that the expected value for 

```
<decision name="list_014" id="_list_014">
        <variable name="list_014"/>
        <literalExpression>
            <text>[1,2,3] in ([1,2,3,4], [1,2,3])</text>
         @</literalExpression>
    </decision> 
```

is correct. Could you explain why is the expected value true? To me it looks like a disjunction of two unary tests where one of the unary tests [1,2,3] in [1,2,3] is false. We probably need to review Table 50 :(

**Comment by StrayAlien** [comment]
Hi @opatrascoiu 

It is discussed here: https://github.com/dmn-tck/tck/pull/204#pullrequestreview-194604731

Greg.

**Comment by opatrascoiu** [comment]
@StrayAlien Thanks for the pointer.

However, I find strange for 

[1,2,3] in ([1,2,3,4], [1,2,3]) to be true
and
[1, 2,3] in ([1, 2, 3]) 
[1,2, 3] in [1, 2, 3]
false

Based on 51.c [1, 2, 3] is a unary test. Hence [1,2, 3] in [1, 2, 3] is equivalent to [1, 2, 3] in =[1, 2, 3], hence the value should be true.

I suggested taking this test out, move forward with merge and raise an RTF to clarify. It seems to me that the change in DMN 1.2 that made tests to be expressions created some ambiguities that need to be clarified.

**Comment by tarilabs** [comment]
I have removed the Approved label based on this comment: https://github.com/dmn-tck/tck/pull/228#issuecomment-487056752

**Comment by StrayAlien** [comment]
Hi @opatrascoiu , @tarilabs.  Thanks.

It feel likes we keep around and around on this one.  These tests are from a PR that was submitted on the 9th December 2018 ... The comments referenced above were on 22nd January.

... we're in May 2019 now.  Just noting!

That aside, I think it actually makes sense.  Think of it like this:

"is [1,2,3] in this list [1,2,3]" no. 

vs

"is [1,2,3] in this list of equality checks : ([1,2,3,4], [1,2,3])'.  Yes.

**Comment by tarilabs** [comment]
> image

> ... we're in May 2019 now. Just noting!

This feels really unfair.

Short recap. We discussed this test case in the original big PR https://github.com/dmn-tck/tck/pull/204#pullrequestreview-194604731 where I raised the same concerns, as for this PR, for a couple of tests suggesting to move them aside (`list014-b`, `list015`) , in order to move on with the bulk of the tests. There was a forward/back on the result of them in this PR https://github.com/dmn-tck/tck/pull/228#issuecomment-462098982, but ultimately those 2 tests were moved aside with commit 121f51f dated 2019-04-06T03:46:09Z and since then this PR is from my perspective is good to go.

Now Octavian has some comments concerning test `list_014` what do you want me to do, just ignore Octavian's feedback?

Here is my pragmatical proposal.

How about moving _also_ `list_014` as subjected of https://github.com/dmn-tck/tck/pull/228#issuecomment-487056752 to a separate PR so we can discuss there and move on with the bulk.

**_As always_**, there are two ways to achieve this: either amend this PR yourself by commenting out also that test, or just let me know and I will append a commit "from backoffice" on this PR to perform the same operation. Just let me know which way is preferred.

Thank you for submitting this test in the first place, but I don't share the same conclusion that "_we're in May 2019 now. Just noting_" happened about these test.

**Comment by opatrascoiu** [comment]
@StrayAlien 

I think your previous statement "is [1,2,3] in this list [1,2,3]" is incorrect. 

Based on production 51.c 
         `comparison = expression, "in", positive unary test`
and the corresponding AST, the second [1, 2, 3] in comparison `[1, 2, 3] in [1, 2, 3]` is a unary test, hence the result of the comparison is true.

The RHS of the `in` expressions is always a unary test.

**Comment by StrayAlien** [comment]
@tarilabs @opatrascoiu - apologies, as I said in a recent post it sometimes very hard not to sound like one being offensive.  Apologies, we're a team here and we're all after the same thing.  I didn't mean to offend.

No, I'd never expect you to merge if there is doubt.  I am happy to comment out that test and raise a new PR after merge as I have done with all the other cases.  I am also happy that you guys cast a critical eye over the tests.  Though (no offense!) I do often wish the process was a smidgeon faster.  I will amend the PR for sure.

@opatrascoiu - as it happens, I did originally have it implemented as you say -  but the consensus in that discussion arrived at the point I mentioned above and I changed the test.  Though I do think it makes sense from @etirelli 's response that `<expr> in <expr>` is different from `<expr> in <unarytests>` - that was what I based my above understanding on.  

I'll remove or comment out that test.  

Greg.

**Comment by opatrascoiu** [comment]
Not sure if I made myself clear. Lets look at this from the AST perspective.

The AST of the `in` expression `[1, 2, 3] in [1, 2, 3]` is built based on production 51.c
comparison = expression, "in", hence the tree is `<expr> in <unarytest>`

The semantics must match the one of unarytests and not general expressions. The tree `<expr> in <expr>` can be obtained if we change production 51.c to comparison = expression, "in", expression

Hope this helps.

**Comment by StrayAlien** [comment]
Thanks @opatrascoiu,

That test is now commented out as discussed.  Re the AST thing - I do understand that - I was making reference to @etirelli's comments re the RHS being an expression vs a unarytests.

But, looking at the grammar, it sure supports what you say about the RHS of an 'in' being a positive unary test (or tests)

```
51. comparison =
<snip>
c. expression , "in" , positive unary test ;
d. expression , "in" , " (", positive unary tests, ")" ;
```
That, along with the following:

image

.. supports that `[1,2,3] in [1,2,3]` should be understood (like you say) as  `[1, 2, 3] in =[1, 2, 3]`, as `<expr> in <unary test>`, not `<expr> in <expr>` as it has been corrected to - at least I _think_ it says that!  To me it just makes more sense and feels a little more 'unified'.  So ... good catch.  Hopefully, we can agree on that.  And, hopefully, it makes more sense to a decision modeller as well.

The code for supporting unary tests doesn't feel that 'unified' with the `not()` ambiguity in there and so on, so every little bit helps that as well. :-)

Additionally, like some free extra steak knives, it might help clear up the other tests that are also commented out.  

Namely, what the boolean result is of the following:

`[1,2,3] in ([[1,2,3,4]], [[1,2,3]])` and `[1,2,3] in ([1,2,3])`.

Greg.
